Automattic/newspack-plugin
# feat(reader-activation): registration block

### All Submissions:






### Changes proposed in this Pull Request:

Implements a simple reader registration block integrated with a reader activation frontend js library. The block features should be later expanded to include:

 - Optional newsletter subscription through API provided by newspack-newsletters with:
   - Default list to assign the reader to, or;
   - Allow the reader to select from a list
 - Social login with Google and Facebook

<img width="1520" alt="image" src="https://user-images.githubusercontent.com/820752/175407621-1915d679-ec50-45c9-94c9-5ad9490f7762.png">

<img width="650" alt="image" src="https://user-images.githubusercontent.com/820752/175407514-3ffd03e2-b50b-4ac8-a2e3-d2f17860aa5e.png">

## Reader Activation Frontend Library

This PR also introduces a small js library to allow easy communication between current and future reader activation front-end components on XHR requests.

Given the optimistic quality of the reader activation's auth intention, once a reader registers through the block, the `readerActivation.setReader( { email } );` is called regardless of authenticating or not. This method dispatches the `reader` event, which can be listened like this:

```js
readerActivation.on( 'reader', ( { detail: { email } } ) => {
	if ( email ) {
		console.log( "Reader updated", email );
	}
} );
```

The library also provides `readerActivation.off( 'reader', cb );` method to remove the listener and `readerActivation.getReader();` to retrieve the currently stored reader data.

In the near future, this library should also be used to update the component that links the user to authenticate or access the reader account page, providing dynamic and appropriate messaging given the authentication state.

## AMP Plus tweak

Similar to what has already been implemented for async/defer, this PR also introduces the support for enabling AMP Plus to a script by using:

```php
\wp_script_add_data( $handle, 'amp-plus', true );
```

This serves as a shortcut to properly add the `data-amp-plus-allowed` attribute to the script.

### How to test the changes in this Pull Request:

1. Confirm you have AMP with AMP Plus enabled and WooCommerce installed for the tests
2. Without `define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );` on your `wp-config.php`, look for the Reader Registration block on a new draft and confirm it's not available, as it shouldn't without reader activation enabled.
3. Add `define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );` and confirm you can add the block
4. Save the page, visit and register an email that doesn't exist on the database
5. Confirm the success message, visit Woo's `my-account` page and confirm you are logged in
6. Log out, register through the block again with the same email
7. Refresh the page and confirm the block does not render, as it identifies you as a registered reader
8. Visit Woo's `my-account` page and confirm you're prompted to authenticate
9. Disable AMP Plus, clear your cookies, register with pure AMP, and confirm you're redirected to a `?newspack_reader=1` page and you're registered (either with auth intention cookie or fully authenticated, depending on the email)

### Other information:
